enhancements/testing/openshift-tests-extension.md

```diff
@@ -0,0 +1,668 @@
+  # as if they are production.
+  "lifecycle": "informing",
+  
+}
+```
+
+##### Run-Test - Running Extension Tests
+The `run-test` verb is used to cause the extension binary to actually run 
+discovered tests. `run-test` accepts test environment arguments as discussed
+in the "list" verb section (in order to reduce fact discovery time needed
+for each test run) and one or more test names to invoke.
+```
+$ extension-binary run-test 
+  --platform      The hardware or cloud platform ("aws", "gcp", "metal", ...).
+  ...other context arguments...
+  --config        The component configuration profile to assume is active on the cluster.
+  --name | -n     Test name to invoke (-n can be specified multiple times).
+```
+
+`run-test` should run the specified tests in parallel. To prevent 
+test cases developing a dependence on one another, `run-test` must make no
+assumption about the number or composition of tests it is asked to run
+(other than they will be consistent with the test environment and isolation
+requirements defined by the "list" verb).
+
+`openshift-tests` will randomize tests into multiple different parallel
+executions of the extension binary in order to amortize initialization
+costs.
+
+Standard output from the invocation will be serialized into JSONL with
+each line associated with a single test run result.
+
+```python
+{ "name": "T1", "result": "success", }
+{ "name": "T1", "result": "success", }
+{ "name": "T1", "result": "success", }
+```
+
+A single test result will contain the following information, encoded on a single line:
+```python
+{
+    "name": "test name",
+    
+    # Duration of test run in milliseconds
+    "duration": "1012",
+    
+    # The outcome of the test; pass, fail, skip, timeout.
+    "result": "pass",
+    "output": "standard output of the test",
+    "error": "standard error of the test",
+    
+    # Human-readable messages to further explain 
+    # skips / timeouts / etc.
+    "messages": [
+        "example message"
+    ]
+}
+```
+
+##### Config - Component Configuration Testing
+
+A component can advertise that it wants to be exercised in multiple different configuration.
+`openshift-tests` will plan an efficient method of testing those configurations and
+ensure tests appropriate to that configuration are run.
+
+The `config` verb will not return until the extension has watched the component
+fully apply the configuration (or it should timeout with an error).
+```
+$ extension-binary config
+  --profile <profile name>      The configuration profile to apply or "default".
+```
+
+The component configuration is considered part of the Environment and thus passed
+in to `list` in order to determine appropriate tests for the configuration. 
+
+Component authors may choose to reduce the number of tests run for non-default
+configuration profiles, focusing only on tests likeliest to fail based on the 
+configuration change, in order to reduce overall execution time.

```

Comment:
Could we get an example on this one, I couldn't come up with a use case immediately.

Reply:
Let's say you want to test a more verbose logging level configuration option in your CRD. You may have hundreds of tests that fully exercise your component in the 'default' configuration, but it is overkill to run them all again simply to verify that debug log statements are being emitted by your pod when verbose logging is enabled.
So you expose an extension configuration for the verbose logging level, and output only one test when asked for a list in that configuration. All that test does is read the component's pod logs and makes sure that it sees debug output.

If you expose a configuration that disables http/1, you might just want to run a test that verifies an http/1 connection is rejected and an http/2 connection is accepted. If you want to test branding, you might just want to verify that the HTML you scrape contains the newly configured name. If you expose a threshold, you might just want to test that a single expected alert is firing after configuring it.

I can add these to the doc if you buy the premise.